Improve .NET-Security section about Antiforgery-Tokens (#212)
OWASP/CheatSheetSeries: cheatsheets/DotNet_Security_Cheat_Sheet.md

### Cross-site request forgery

DO NOT: Send sensitive data without validating Anti-Forgery-Tokens.

DO: Send the anti-forgery token with every POST/PUT request:

> #### Using the full .NET-Framework :

Reviewer: ```suggestion
#### Using .NET Framework:
```

> DO NOT: Send sensitive data without validating Anti-Forgery-Tokens ([.NET](https://docs.microsoft.com/en-us/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks) / [.NET Core](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration)).

Reviewer 2: Better to link to the published MS documentation rather than the GitHub:

https://docs.microsoft.com/en-us/aspnet/core/security/anti-request-forgery?view=aspnetcore-3.0#aspnet-core-antiforgery-configuration

Author: Fixed.

}
```

#### Using .NET Core 2.0 or later:

After .NET Core 2.0 it is possible to [automatically generate and verify the antiforgery token](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration).

If you are using [tag-helpers](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/intro) - which is the default for most web project templates - all forms will automatically send the anti-forgery token.
You can check if tag-helpers are enabled by checking if your main `_ViewImports.cshtml` file contains:

```csharp
@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
```

`IHtmlHelper.BeginForm` also sends anti-forgery-tokens automatically.

If you are not using tag-helpers or `IHtmlHelper.BeginForm`, forms must have the requisite helper as seen here:

```html
<form action="RelevantAction" >
@Html.AntiForgeryToken()
</form>
```

> And then add the `[AutoValidateAntiforgeryToken]` attribute to the action method. This will validate the token for all HTTP methods other than GET, HEAD, OPTIONS and TRACE.

Author: Good point - I reworked the section to reflect this. What do you think @ThunderSon ?

}
```

#### Using .NET Core 2.0 or later:

> After .NET Core 2.0 it is possible to [automatically generate and verify the antiforgery token](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration).

Reviewer 3: `After` here is ambiguous. You might mean `Starting with`...

Author: Thanks for the feedback, I learned something as a non-native speaker ;) I changed it.

> If you are using [tag-helpers](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/intro) - which is the default for most web project templates - all forms will automatically send the anti-forgery token.

Reviewer 3: I'd probably use `,`s instead of `-`s. And I'd probably include a `then` before `all`...

Author: Done.

You can check if tag-helpers are enabled by checking if your main `_ViewImports.cshtml` file contains:

```csharp
@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
```

`IHtmlHelper.BeginForm` also sends anti-forgery-tokens automatically.

> If you are not using tag-helpers or `IHtmlHelper.BeginForm`, forms must have the requisite helper as seen here:

Reviewer 3: I'd probably include a `then` before `forms`.

Personally, I'm not a fan of `not using ... or`. I'd write something like `if you are using neither ... nor ...`, but people don't really like that. The alternative is `Unless you are using X or Y, you must ...` which is probably better.

Author: Done.

```html
<form action="RelevantAction" >
@Html.AntiForgeryToken()
</form>
```

> To automatically validate all requests other than GET, HEAD, OPTIONS and TRACE you need to add a global action filter with the [AutoValidateAntiforgeryToken](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.autovalidateantiforgerytokenattribute?view=aspnetcore-2.2) attribute inside your `Startup.cs` ([Source](https://andrewlock.net/automatically-validating-anti-forgery-tokens-in-asp-net-core-with-the-autovalidateantiforgerytokenattribute/)):

Reviewer: ```suggestion
To automatically validate all requests other than GET, HEAD, OPTIONS and TRACE you need to add a global action filter with the [AutoValidateAntiforgeryToken](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.autovalidateantiforgerytokenattribute?view=aspnetcore-2.2) attribute inside your `Startup.cs` as mentioned in the following [article](https://andrewlock.net/automatically-validating-anti-forgery-tokens-in-asp-net-core-with-the-autovalidateantiforgerytokenattribute/):
```

}
```

#### Using .NET Core 2.0 or later:

> Starting with .NET Core 2.0 it is possible to [automatically generate and verify the antiforgery token](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration).

Reviewer 2: As above, link to published documents

Author: Fixed.

If you are using [tag-helpers](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/intro), which is the default for most web project templates, then all forms will automatically send the anti-forgery token.

> You can check if tag-helpers are enabled by checking if your main `_ViewImports.cshtml` file contains:

Reviewer 2: Single newlines like this get ignored by markdown, so can be removed. This and the line above just show as one sentence.

Author: Fixed.

}
```

#### Using .NET Core 2.0 or later:

After .NET Core 2.0 it is possible to automatically generate and verify the antiforgery token.

> If you are using tag-helpers - what is the default for most web project templates - all forms will automatically send the anti-forgery token.

Reviewer 2: `what it` should be `which is`

Author: Fixed.

> If you are using tag-helpers - which is the default for most web project templates - all forms will automatically send the anti-forgery token.

Reviewer: Kindly link tag-helpers to the following [doc](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/intro)

> You can check if tag-helpers are enabled by checking if your main `_ViewImports.cshtml`- file contains:

Reviewer 2: There's an extra hyphen after the filename

Author: I removed it.

```csharp
@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
```

> `IHtmlHelper.BeginForm` also sends anti-forgery-tokens automatically.

Reviewer: Can you add a link to where that is mentioned, so that the reader is able to properly understand what's going underneath if they need to?

Author: I added the link where it is mentioned to the beginning of the parent section and this subsection, as it describes TagHelpers as well as the BeginForm-topic.

OK @ThunderSon ?
https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration
`Similarly, IHtmlHelper.BeginForm generates antiforgery tokens by default if the form's method isn't GET.`

Reviewer: Yes, makes sense. Sadly, the link to `IHtmlGelper.BeginForm` is broken in their docs

If you are not using tag-helpers or `IHtmlHelper.BeginForm`, forms must have the requisite helper as seen here:

```html
<form action="RelevantAction" >
@Html.AntiForgeryToken()
</form>
```

And then add the `[AutoValidateAntiforgeryToken]` attribute to the action method. This will validate the token for all HTTP methods other than GET, HEAD, OPTIONS and TRACE.

> #### Using .Net Core 2.0 or full .NET-Framework with AJAX

Reviewer: ```suggestion
#### Using .Net Core 2.0 or .NET Framework with AJAX
```

> You will need to attach the anti-forgery token to Ajax requests.

Reviewer 2: A couple of lines above this you have `AJAX`, but here's it's `Ajax`.  I've seen people use both, but `AJAX` is more widely used in the other cheatsheets, so could you change this one for consistency?

Author: Makes sense - done.

> If you are using jQuery in an ASP .NET Core MVC view this can be achieved using this snippet:

Reviewer: ASP.NET ; attach it so it becomes `ASP.NET` instead of `ASP .NET`

```javascript
@inject  Microsoft.AspNetCore.Antiforgery.IAntiforgery antiforgeryProvider
$.ajax(
{
    type: "POST",
    url: '@Url.Action("Action", "Controller")',
    contentType: "application/x-www-form-urlencoded; charset=utf-8",
    data: {
        id: id,
        '__RequestVerificationToken': '@antiforgeryProvider.GetAndStoreTokens(this.Context).RequestToken'
    }
})
```

> If you are using the full .NET-Framework you can find some snippets [here](https://docs.microsoft.com/de-de/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks#anti-csrf-and-ajax).

Reviewer 2: Change this to the English version (`en-us` rather than `de-de` in the URL).

Author: Fixed.

> If you are using the full .NET-Framework you can find some snippets [here](https://docs.microsoft.com/en-us/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks#anti-csrf-and-ajax).

Reviewer: ```suggestion
If you are using the .NET Framework, you can find some code snippets [here](https://docs.microsoft.com/en-us/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks#anti-csrf-and-ajax).
```

More information can be found [here](Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md) for Cross-Site Request Forgery.

> More information about Anti-Forgery-Tokens can be found [here](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration) for .NET Core and [here](https://docs.microsoft.com/de-de/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks) for the full .NET-Framework.

Reviewer 2: Again, please update to English

Author: Fixed.

> More information about Anti-Forgery-Tokens can be found [here](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration) for .NET Core and [here](https://docs.microsoft.com/en-us/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks) for the full .NET-Framework.

Reviewer: Mention it as an inline url on [line 566](https://github.com/OWASP/CheatSheetSeries/pull/212/files#diff-00bdcd96a1637f6a84a020c2a6365b9fR566) and remove this line
